src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java

```diff
@@ -99,24 +105,88 @@ public boolean isJava8() {
             && getRuntime().startsWith("java8");
   }
 
-  private static String getTopLevelValue(Document doc, String parentTagName, String childTagName) {
-    try {
-      NodeList parentElements = doc.getElementsByTagNameNS(APP_ENGINE_NAMESPACE, parentTagName);
-      if (parentElements.getLength() > 0) {
-        Node parent = parentElements.item(0);
-        if (parent.hasChildNodes()) {
-          for (int i = 0; i < parent.getChildNodes().getLength(); ++i) {
-            Node child = parent.getChildNodes().item(i);
-            if (child.getNodeName().equals(childTagName)) {
-              return child.getTextContent();
+  /**
+   * Given the following structure:
+   * <pre>
+   * {@code
+   * <env-variables>
+   *   <env-var name="key" value="val" />
+   * </env-variables>
+   * }
+   * </pre>
+   * This will construct a map of the form {[key, val], ...}
+   *
+   * @return a map representing the environment variable settings in the appengine-web.xml.
+   */
+  public Map<String, String> getEnvironment() {
+    Node environmentParentNode = getNode(document, "appengine-web-app", "env-variables");
+    return getAttributeMap(environmentParentNode, "env-var", "name", "value");
+  }
+
+  private static String getText(Node node) {
+    if (node != null) {
+      try {
+        return node.getTextContent();
+      } catch (DOMException ex) {
+        // this shouldn't happen barring a very funky DOM implementation
```

Comment:
maybe. the question is, if the parsing fails, do we want to throw an exception, or fall back on default values (how it was working before)? I'm in favor of runtime exception.